src/Arcus.Messaging.Pumps.ServiceBus/ServiceBusMessageRouter.cs

```diff
@@ -0,0 +1,197 @@
+
+                        hasGoneThroughMessageHandler = true;
+                        if (isProcessed)
+                        {
+                            await PotentiallyAutoCompleteMessageAsync(message, messageContext);
+                            return MessageProcessingResult.Success(message.MessageId);
+                        }
+                    }
+                }
+
+                if (hasGoneThroughMessageHandler)
+                {
+                    await AbandonMessageMatchedHandlerFailedAsync(messageContext);
+                    return MessageProcessingResult.Failure(message.MessageId, MatchedHandlerFailed, "Failed to process Azure Service Bus message in pump as the matched handler did not successfully processed the message");
+                }
+
+                await DeadLetterMessageNoHandlerMatchedAsync(messageContext);
+                return MessageProcessingResult.Failure(message.MessageId, CannotFindMatchedHandler, "Failed to process Azure Service Bus message in pump as no message handler was matched against the message");
+            }
+            catch (Exception exception)
+            {
+                Logger.LogCritical(exception, "Unable to process message with ID '{MessageId}'", message.MessageId);
+
+                await messageContext.AbandonMessageAsync(new Dictionary<string, object>(), CancellationToken.None);
+                return MessageProcessingResult.Failure(message.MessageId, ProcessingInterrupted, "Failed to process message in pump as there was an unexpected critical problem during processing, please see the logs for more information", exception);
+            }
+        }
+
+        private static string LoadMessageBody(ServiceBusReceivedMessage message, AzureServiceBusMessageContext context)
+        {
+            Encoding encoding = DetermineEncoding();
+            string messageBody = encoding.GetString(message.Body.ToArray());
+
+            return messageBody;
+
+            Encoding DetermineEncoding()
+            {
+                Encoding fallbackEncoding = Encoding.UTF8;
+
+                if (context.Properties.TryGetValue(PropertyNames.Encoding, out object encodingNameObj)
+                    && encodingNameObj is string encodingName
+                    && !string.IsNullOrWhiteSpace(encodingName))
+                {
+                    EncodingInfo foundEncoding =
+                        Encoding.GetEncodings()
+                                .FirstOrDefault(e => e.Name.Equals(encodingName, StringComparison.OrdinalIgnoreCase));
+
+                    return foundEncoding?.GetEncoding() ?? fallbackEncoding;
+                }
+
+                return fallbackEncoding;
+            }
+        }
+
+        private async Task PotentiallyAutoCompleteMessageAsync(ServiceBusReceivedMessage message, AzureServiceBusMessageContext messageContext)
```

Comment:
I believe we can remove the `message` parameter ?  Isn't everything available through he `messageContext` parameter ?

Reply:
Done here: https://github.com/arcus-azure/arcus.messaging/pull/635/commits/69827bd7dc6de6f3347fe1e1794374b91be75ece